programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs

```diff
@@ -0,0 +1,128 @@
+    #[account(mut)]
+    pub user: Signer<'info>,
+
+    #[account(
+        seeds = [b"perpetuals"],
+        bump = perpetuals.perpetuals_bump
+    )]
+    pub perpetuals: Box<Account<'info, Perpetuals>>,
+
+    #[account(
+        seeds = [b"pool",
+                 pool.name.as_bytes()],
+        bump = pool.bump
+    )]
+    pub pool: Box<Account<'info, Pool>>,
+
+    #[account(
+        seeds = [b"custody",
+                 pool.key().as_ref(),
+                 custody.mint.as_ref()],
+        bump = custody.bump
+    )]
+    pub custody: Box<Account<'info, Custody>>,
+
+    #[account(
+        // Custom oracle must first be initialized by authority before permissionless updates.
+        mut,
+        seeds = [b"oracle_account",
+                 pool.key().as_ref(),
+                 custody.mint.as_ref()],
+        bump
+    )]
+    pub oracle_account: Box<Account<'info, CustomOracle>>,
+
+    /// CHECK: Needed for ed25519 signature verification, to inspect all instructions in this transaction.
+    #[account(address = IX_ID)]
+    pub ix_sysvar: AccountInfo<'info>,
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize, Copy, Clone)]
+pub struct SetCustomOraclePricePermissionlessParams {
+    pub price: u64,
+    pub expo: i32,
+    pub conf: u64,
+    pub ema: u64,
+    pub publish_time: i64,
+}
+
+pub fn set_custom_oracle_price_permissionless<'info>(
+    ctx: Context<'_, '_, '_, 'info, SetCustomOraclePricePermissionless<'info>>,
+    params: &SetCustomOraclePricePermissionlessParams,
+) -> Result<()> {
+    // If the incoming publish time is behind or equal to the current, then no-op.
+    if params.publish_time <= ctx.accounts.oracle_account.publish_time {
+        msg!("Publish time is behind the current, so CustomOracle price did not update.");
+        return Ok(());
+    }
+
+    ctx.accounts.oracle_account.set(
+        params.price,
+        params.expo,
+        params.conf,
+        params.ema,
+        params.publish_time,
+    );
+    // Get what should be the Ed25519Program signature verification instruction.
+    let signature_ix: Instruction = load_instruction_at_checked(0, &ctx.accounts.ix_sysvar)?;
+    validate_ed25519_signature_instruction(
+        &signature_ix,
+        &ctx.accounts.custody.oracle.permissionless_update_pubkey,
+        &ctx.accounts.custody.key(),
+        &ctx.accounts.oracle_account.serialize(),
+    )
+}
+
+fn validate_ed25519_signature_instruction(
+    signature_ix: &Instruction,
+    expected_pubkey: &Pubkey,
+    custody_key: &Pubkey,
+    serialized_oracle: &[u8],
+) -> Result<()> {
+    require_eq!(
+        signature_ix.program_id,
+        ED25519_ID,
+        PerpetualsError::PermissionlessOraclePriceSignatureFailure
+    );
+    require_eq!(
+        signature_ix.accounts.len(),
+        0,
+        PerpetualsError::PermissionlessOraclePriceSignatureFailure
+    );
+
+    let verified_pubkey = &signature_ix.data[16..16 + 32];
+    let verified_message = &signature_ix.data[112..];
```

Comment:
all this math is unclear, could you add a comment and explicitly check for data size before slicing

Reply:
yes!  added more checks above for the message data size, and also added comment for where the offsets came from